Docs: fix GraphQL Reference (#32543)
gatsbyjs/gatsby: docs/docs/graphql-reference.md

        version
        packageJson {
          description
        }
      }
    }
  }
}
```

In the GraphiQL editor, try using the editor's autocomplete (`Ctrl + Space`) to get extended details from the `packageJson` nodes.

If you're using Gatsby version `2.2.0` or later, you can remove `edges` and `node` from your query and replace it with `nodes`. The query will still work and the returned object will reflect the `nodes` structure.

```graphql
{
  allSitePlugin {
    totalCount
    nodes {
      name
      version
      packageJson {
        description
      }
    }
  }
}
```

## Limit

There are several ways to reduce the number of results from a query. Here `totalCount` tells you there's 8 results, but `limit` is used to show only the first three.

```graphql
{
  allMarkdownRemark(limit: 3) {
    totalCount
    edges {
      node {
        frontmatter {
          title
        }
      }
    }
  }
}
```

## Skip

Skip over a number of results. In this query `skip` is used to omit the first 3 results.

```graphql
{
  allMarkdownRemark(skip: 3) {
    totalCount
    edges {
      node {
        frontmatter {
          title
        }
      }
    }
  }
}
```

## Filter

> In this query `filter` and the `ne` (not equals) operator is used to show only results that have a title. You can find a good video tutorial on this [here](https://www.youtube.com/watch?v=Lg1bom99uGM).

Reviewer: ```suggestion
In this query the `filter` and `ne` (not equals) operators are used to show only results that have a title. You can find a good video tutorial on this [here](https://www.youtube.com/watch?v=Lg1bom99uGM).
```

Might be a non-issue (might be just me and 🇩🇪)?!

Author: I read this as: "the [modifier] operator" – i.e. "filter" is not an operator

Reviewer: Ah, that makes sense of course!

```graphql
{
  allMarkdownRemark(filter: { frontmatter: { title: { ne: "" } } }) {
    totalCount
    edges {
      node {
        frontmatter {
          title
        }
      }
    }
  }
}
```

> Gatsby relies on [Sift](https://www.npmjs.com/package/sift) to enable MongoDB-like query syntax for object filtering. This allows Gatsby to support operators like `eq`, `ne`, `in`, `regex` and querying nested fields through the `__` connector.

Reviewer: General comment:

* we're talking about the "`__` connector` here, but don't offer an example for it
* the first example I'm seeing is in the "Sort" section (https://github.com/gatsbyjs/gatsby/blob/f7ab2e64aea104f39ec9c3e2be636f5bdd0b5d53/docs/docs/graphql-reference.md#sort), and the connector there is `___` – three underscores instead of the two shown here

I'm a GraphQL noob, and out of touch with allthesethings anyway, so maybe I'm just confusing two different things?!

Author: Pretty sure tis is a typo – good catch!

It is also possible to filter on multiple fields by separating the individual filters by a comma (works as an AND):

```graphql
filter: { contentType: { in: ["post", "page"] }, draft: { eq: false } }
```

> In this query the fields `categories` and `title` are filtered to find the book that has `Fantastic` in its title and belongs to the `magical creatures` category.

Reviewer: ```suggestion
In this query the fields `categories` and `title` are filtered to find the book that has `Fantastic` in its title _and_ belongs to the `magical creatures` category.
```

Reviewer: (Trying to emphasize the `AND`, and connecting it to the preceding paragraph/mention.)

```suggestion
In this query the fields `categories` and `title` are filtered to find the book that has `Fantastic` in its title _AND_ belongs to the `magical creatures` category.
```

> In this query the fields `categories` and `title` are filtered to find the book that has `Fantastic` in its title and belongs to the `magical creatures` category.

Reviewer: ```suggestion
In this query the fields `categories` and `title` are filtered to find the book that belongs to the `magical creatures` category _AND_ has `Fantastic` in its title.
```

* switch order to match query order
* emphasize `AND` (picking up its first mention in the preceding paragraph)

- `lte`: short for **less than or equal**, must be less than or equal to given value
- `elemMatch`: short for **element match**, this indicates that the field you are filtering will return an array of elements, on which you can apply a filter using the previous operators

```graphql
{
  # eq: I want all the titles that match "Fantastic Beasts and Where to Find Them"
  example_eq: allMarkdownRemark(
    filter: {
      frontmatter: { title: { eq: "Fantastic Beasts and Where to Find Them" } }
    }
  ) {
    edges {
      node {
        frontmatter {
          title
        }
      }
    }
  }

  # neq: I want all the titles which are NOT equal to the empty string
  example_ne: allMarkdownRemark(
    filter: { frontmatter: { title: { ne: "" } } }
  ) {
    edges {
      node {
        frontmatter {
          title
        }
      }
    }
  }

  # regex: I want all the titles that do not start with 'T' -- this is what /^[^T]/ means.

>   # To learn more about regular expressions: https://regexr.com/

Reviewer: <3 regexr

By default, sort `fields` will be sorted in ascending order. Optionally, you can specify a sort `order` per field by providing an array of `ASC` (for ascending) or `DESC` (for descending) values. For example, to sort by `frontmatter.date` in ascending order, and additionally by `frontmatter.title` in descending order, you would use `sort: { fields: [frontmatter___date, frontmatter___title], order: [ASC, DESC] }`. Note that if you only provide a single sort `order` value, this will affect the first sort field only, the rest will be sorted in default ascending order.

```graphql
{
  allMarkdownRemark(
    sort: {
      fields: [frontmatter___date, frontmatter___title]
      order: [ASC, DESC]
    }
  ) {
    totalCount
    edges {
      node {
        frontmatter {
          title
          date
        }
      }
    }
  }
}
```

## Format

### Dates

Dates can be formatted using the `formatString` function.

```graphql
{
  allMarkdownRemark(filter: { frontmatter: { date: { ne: null } } }) {
    edges {
      node {
        frontmatter {
          title
          date(formatString: "dddd DD MMMM YYYY")
        }
      }
    }
  }
}
```

> Gatsby relies on [Moment.js](https://momentjs.com/) to format the dates. This allows you to use any tokens in your string. See [moment.js documentation](https://momentjs.com/docs/#/displaying/format/) for more tokens.

Reviewer: `Moment.js` likes the uppercase first char when used as a product name (ref. https://momentjs.com) – if I get this right, we want to be consistent here:

```suggestion
Gatsby relies on [Moment.js](https://momentjs.com/) to format the dates. This allows you to use any tokens in your string. See the [Moment.js documentation](https://momentjs.com/docs/#/displaying/format/) for more tokens.
```

To add variables to page component queries, pass these in the `context` object [when creating pages](/docs/creating-and-modifying-pages/#creating-pages-in-gatsby-nodejs).

```graphql
query GetBlogPosts(
  $limit: Int
  $filter: MarkdownRemarkFilterInput
  $sort: MarkdownRemarkSortInput
) {
  allMarkdownRemark(limit: $limit, filter: $filter, sort: $sort) {
    edges {
      node {
        frontmatter {
          title
          date(formatString: "dddd DD MMMM YYYY")
        }
      }
    }
  }
}
```

> ```graphql

Reviewer: Wondering if including `MarkdownRemarkFilterInput` and `[…]SortInput`, or even `createPage({ … })` would be … longer but clearer?
